packages/nouns-contracts/contracts/governance/ExcessETH.sol

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: GPL-3.0
+
+/// @title A helpder contract for calculating Nouns excess ETH
+
+/**
+ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░█████████░░█████████░░░ *
+ * ░░░░░░██░░░████░░██░░░████░░░ *
+ * ░░██████░░░████████░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░░░░░█████████░░█████████░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ *
+ */
+
+pragma solidity ^0.8.19;
+
+import { IExcessETH } from '../interfaces/IExcessETH.sol';
+import { Ownable } from '@openzeppelin/contracts/access/Ownable.sol';
+import { IERC20 } from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import { INounsAuctionHouse } from '../interfaces/INounsAuctionHouse.sol';
+
+interface RocketETH {
+    function getEthValue(uint256 _rethAmount) external view returns (uint256);
+}
+
+interface INounsDAOV3 {
+    function adjustedTotalSupply() external view returns (uint256);
+}
+
+interface INounsAuctionHouseV2 is INounsAuctionHouse {
+    function prices(uint256 auctionCount) external view returns (Settlement[] memory settlements);
+}
+
+contract ExcessETH is IExcessETH, Ownable {
+    /**
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     *   ERRORS
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    error NotEnoughAuctionHistory();
+
+    error RocketETHConversionRateTooLow();
+    /**
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     *   EVENTS
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    event AuctionSet(address oldAuction, address newAuction);
+    event NumberOfPastAuctionsForMeanPriceSet(
+        uint16 oldNumberOfPastAuctionsForMeanPrice,
+        uint16 newNumberOfPastAuctionsForMeanPrice
+    );
+
+    /**
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     *   STATE
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    INounsDAOV3 public immutable dao;
+    INounsAuctionHouseV2 public immutable auction;
+    IERC20 public immutable wETH;
+    IERC20 public immutable stETH;
+    IERC20 public immutable rETH;
+    uint16 public numberOfPastAuctionsForMeanPrice;
+
+    constructor(
+        address owner_,
+        INounsDAOV3 dao_,
+        INounsAuctionHouseV2 auction_,
+        IERC20 wETH_,
+        IERC20 stETH_,
+        IERC20 rETH_,
+        uint16 numberOfPastAuctionsForMeanPrice_
+    ) {
+        _transferOwnership(owner_);
+
+        dao = dao_;
+        auction = auction_;
+        wETH = wETH_;
+        stETH = stETH_;
+        rETH = rETH_;
+        numberOfPastAuctionsForMeanPrice = numberOfPastAuctionsForMeanPrice_;
+    }
+
+    /**
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     *   PUBLIC FUNCTIONS
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    function excessETH() public view returns (uint256) {
+        uint256 expectedTreasuryValue = meanAuctionPrice() * dao.adjustedTotalSupply();
+        return min(treasuryValueInETH() - expectedTreasuryValue, owner().balance);
```

Comment:
Perhaps throw a custom error instead if there is there is no excess ETH.

Reply:
this min is meant to allow us to burn as much as we can; eg excess is 3K ETH but we only have 2K in native ETH, let's at least burn that.

for the case of no excess I agree we should just not burn, will add a condition for that in the treasury contract